R/grid_helpers.R

```diff
@@ -319,27 +319,46 @@ compute_grid_info <- function(workflow, grid) {
   res <- min_grid(extract_spec_parsnip(workflow), grid)
 
   if (any_parameters_preprocessor) {
-    res$.iter_preprocessor <- seq_len(nrow(res))
+    res$.iter_preprocessor <-
+      vctrs::vec_group_id(res[parameters_preprocessor$id])
+    attr(res$.iter_preprocessor, "n") <- NULL
+    model_iters_needed <- duplicated(res$.iter_preprocessor)
   } else {
     res$.iter_preprocessor <- 1L
+    model_iters_needed <- TRUE
   }
 
   res$.msg_preprocessor <-
     new_msgs_preprocessor(
-      seq_len(max(res$.iter_preprocessor)),
+      res$.iter_preprocessor,
       max(res$.iter_preprocessor)
     )
 
-  if (nrow(res) != nrow(grid) ||
-      (any_parameters_model && !any_parameters_preprocessor)) {
-    res$.iter_model <- seq_len(dplyr::n_distinct(res[parameters_model$id]))
-  } else {
-    res$.iter_model <- 1L
+  res$.iter_model <- 1L
+
+  if (any_parameters_model) {
+    model_iter_ids <- vctrs::vec_group_id(
+      res[model_iters_needed, parameters_model$id]
+    )
+    res$.iter_model[model_iters_needed] <- model_iter_ids + 1
+  }
+
+  if (isTRUE(model_iters_needed)) {
+    res$.iter_model <- res$.iter_model - 1
   }
```

Comment:
Not super happy with this pattern—feels like I'm missing a much simpler off-by-one fix?